feat: Make smock work with the EDR-powered version of Hardhat

fvictorio opened the pull request:
We are going to release a new version of Hardhat this week that significantly changes its internals. Since Smock relies on several internal APIs, it will stop working with it. This PR makes the necessary changes for it to work.

This version of Smock _won't_ work with previous versions of Hardhat. We've updated the peer dependency accordingly. Adding this change in a way that supported both new and old versions of Hardhat was possible, but too complex for the time we had to work on this.

The main change here is that, instead of taking a rxjs stream of call results and modifying their values, this version uses a callback that is executed for each call response. The callback can return `undefined` to leave the answer as is, or return a modified version of it. The sandbox uses the information about the call (the recipient address and the calldata) to figure out which mocked function to "ask" for the mocked answer.

0xOneTony commented on src/logic/programmable-function-logic.ts:
```diff
@@ -26,20 +27,16 @@ export class ProgrammableFunctionLogic extends WatchableFunctionLogic {
   protected answerByArgs: { answer: ProgrammedAnswer; args: unknown[] }[] = [];
 
   constructor(
+    private address: Address,
```
I belive this address is not being used

fvictorio replied:
Good catch, thanks! Fixed in c4177d30174c731ad162148ce94a82221cbe0d22

0xOneTony commented on src/logic/programmable-function-logic.ts:
```diff
@@ -82,40 +79,43 @@ export class ProgrammableFunctionLogic extends WatchableFunctionLogic {
 
-  private async modifyAnswer(result: EVMResult, call: ContractCall): Promise<void> {
-    const answer = this.getCallAnswer(call);
+  async getEncodedCallAnswer(address: Address, data: Buffer): Promise<[result: Buffer, shouldRevert: boolean] | undefined> {
```
Here as well, address is not used

0xOneTony commented on src/logic/programmable-function-logic.ts:
```diff
@@ -82,40 +79,43 @@ export class ProgrammableFunctionLogic extends WatchableFunctionLogic {
-        result.execResult.returnValue = await this.encodeValue(answer.value, call);
+        return [this.encodeRevertReason(answer.value), answer.shouldRevert];
+      }
+      {
+        return [await this.encodeValue(answer.value, data), answer.shouldRevert];
       }
```
```suggestion
      return [await this.encodeValue(answer.value, data), answer.shouldRevert];
```

0xOneTony reviewed (commented):
This is great, thank you! 🔥 You mentioned this won't work with previous versions of Hardhat, maybe we should also disclose this in the README.

fvictorio commented:
Thanks a lot for reviewing this @0xOneTony!

I think I applied all your suggestions, and I also added a note in the readme. Let me know if you'd rather have that in a different place or with a different phrasing.

I also updated typescript because I was getting a weird compilation error, which I think was related to bumping the version of `@types/node`.

0xOneTony reviewed (dismissed):
LGTM! 🚀
